In [1]:
# The code was removed by Watson Studio for sharing.

# Segmenting and Clustering Neighborhoods in Toronto

_Author: Miguel Acevedo_

On this notebook I explore, segment, and cluster the neighborhoods in the city of Toronto. For the Toronto neighborhood data, I scrape the [Wikipedia page](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) containing the Zip code, Borough and Neighborhood data for each neighborhood in Toronto, wrangle, clean and read it into a Pandas Dataframe.   

I then convert these addresses into their equivalent latitude and longitude values using [geopy library](https://geopy.readthedocs.io/en/latest/) and use the [Foursquare Places API](https://developer.foursquare.com/docs/places-api/) to get the most common venues categories in each neigborhood to group them into clusters using K-Means.

***

En este Notebook exploro, segmento y agrupo los vecindarios de la ciudad de Toronto. Para obtener los datos del vecindario de Toronto, hago _webscraping_ de la [página de Wikipedia](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) que contiene el código postal, la ciudad y los datos del vecindario para cada vecindario en Toronto. Una vez obtengo los datos entonces los limpio y lo leo en un Pandas Dataframe.

Luego, convierto estas direcciones en sus valores equivalentes de latitud y longitud y utilizo la API de Foursquare para obtener las categorías de lugares más comunes en cada vecindario y asi agruparlas en grupos similares utilizando el algoritmo de K-Means.

***

## Table of contents
> 1. [Download and Explore Data](#download)
> 2. [Explore Neighborhoods in Toronto](#explore)
> 3. [Analyze each Neighborhood](#analyze)
> 4. [Cluster Neighborhoods](#cluster)
> 5. [Examine Clusters](#examine)

***

## Importing Libraries

In [2]:
import numpy as np # handle data into vectorized manner
import pandas as pd # data analysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # handle json files
from pandas.io.json import json_normalize # transform JSON file to pandas dataframe
import requests # handle requests
from bs4 import BeautifulSoup
import csv

!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# Clustering Model
from sklearn.cluster import KMeans

!pip install folium==0.5.0
import folium # map rendering library

print('Libraries imported!')

     |████████████████████████████████| 81kB 7.8MB/s eta 0:00:011
  Stored in directory: /home/dsxuser/.cache/pip/wheels/f8/98/ff/954791afc47740d554f0d9e5885fa09dd60c2265d42578e665
Successfully built folium
Libraries imported!


## <a name="download"></a>Download and Explore Data

***

The following code makes a request to the Wikipedia site, parses data using Beautiful Soup and saves each row into a csv file:  
**This function was deprecated. Please see note below**

In [3]:
# get html using requests and parse with Beautiful Soup
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' 


def wikepedia_scraper(url):
    """ 
    This function takes in the URL as an input, gets html, parses website using beatiful soup,
    prints out the results for manual inspection, then it outputs a csv file with the data. 
    """
    
    r = requests.get(url, timeout=3).text # get html 
    soup = BeautifulSoup(r, 'lxml') # parse html with Beautiful Soup
    table = soup.table # go to table

    # create csv file to save data:
    csv_file = open("toronto_neighborhoods.csv", "w")
    csv_writer = csv.writer(csv_file)
    csv_writer.writerow(['PostalCode', 'Borough', 'Neighborhood'])


    # loop through each row and scrape data:
    for row in table.find_all('td'): # find all rows within table
        try: 
            zip_code = row.p.b.text # scrape zip code
            borough = row.p.span.a.text.strip() # scrape borough
            neighborhood = row.span.text.split('(')
            neighborhood = neighborhood[1].split(')')[0]
            if '/' in neighborhood: # If multiple Neighborhoods, separate with comma
                neighborhood = neighborhood.replace(' /', ',')
                print(zip_code)
                print(borough)
                print(neighborhood)
                print()
            else:
                neighborhood = neighborhood
                print(zip_code)
                print(borough)
                print(neighborhood)
                print()
        except Exception as e: 
            zip_code = None
            borough = None
            neighborhood = None
        print() # print results
    
        csv_writer.writerow([zip_code, borough, neighborhood]) # Save to csv file
    
    csv_file.close() # close csv file

# wikepedia_scraper(url)
# df = pd.read_csv('toronto_neighborhoods.csv')
# df.head()

> **Note:** The structure of the webpage changed so the above function would not work anymore.
> I left the function on the notebook to ouline how I would scrape the webpage with the old structure.  

***

Since the data is now on a simpler format, I used Pandas library to scrape the table as shown below:

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)[0] # use pandas to scrape table. 
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

Dropping the zip codes that are not assigned:

In [5]:
df = df.loc[df['Borough'] != 'Not assigned', :].reset_index(drop=True)
df.head()

Postal code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [6]:
func = lambda x: x.replace(' /', ',') # lambda function that separates Neighborhoods with a comma
df['Neighborhood'] = df['Neighborhood'].apply(func)
df.head(10)

Postal code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke                             Islington Avenue
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

In [7]:
print(df.shape)
df.head()

(103, 3)


Postal code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Geolocation data

In [8]:
# combine Postal code, Borough and Neighborhood with state and country to get geolocation
x = []
for i in df.index.values:
  x.append(f"{df['Borough'][i]}, {df['Postal code'][i]}, Toronto, Canada")
df['Location'] = x

# use geopy Pandas Dataframe option to get geolocation for each row
from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent="toronto_location") # instantiate geolocator

from geopy.extra.rate_limiter import RateLimiter
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1) # call geocoding service
df['location2'] = df['Location'].apply(geocode)
df['Latitude'] = df['location2'].apply(lambda loc: tuple(loc.point)[0] if loc else None) # get latitude
df['Longitude'] = df['location2'].apply(lambda loc: tuple(loc.point)[1] if loc else None) # get longitude
df.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

                                 Location  \
0        North York, M3A, Toronto, Canada   
1        North York, M4A, Toronto, Canada   
2  Downtown Toronto, M5A, Toronto, Canada   
3        North York, M6A, Toronto, Canada   
4  Downtown Toronto, M7A, Toronto, Canada   

                                           location2   Latitude  Longitude  
0  (North York, Toronto, Golden Horseshoe, Ontari...  43.754326 -79.449117  
1                                               None        NaN        NaN  
2                                               None        NaN        NaN  
3                                               None        NaN        NaN  
4  (Downtown Yonge, Toronto Centre, Old Toronto, ...  43.656322 -79.380916

In [9]:
df['location2'].count()

17

**Unfortunately, the Geopy module is inconsistent and was only able to find Latitude and Longitude values for 18 out of the 103 Boroughs. I will use the Google [Geocoding API](https://developers.google.com/maps/documentation/geocoding/start?hl=es_419) python library to convert those to geolocation.**

### Using Google Geolocation API

**Install and instantiate Google Maps Geocoding client:**

In [12]:
# The code was removed by Watson Studio for sharing.

Requirement already up-to-date: googlemaps in /opt/conda/envs/Python36/lib/python3.6/site-packages (4.2.0)


**Testing:**

In [13]:
# test
geocode_result = gmaps.geocode('1600 Amphitheatre Parkway, Mountain View, CA')
print(geocode_result)

[{'address_components': [{'long_name': '1600', 'short_name': '1600', 'types': ['street_number']}, {'long_name': 'Amphitheatre Parkway', 'short_name': 'Amphitheatre Pkwy', 'types': ['route']}, {'long_name': 'Mountain View', 'short_name': 'Mountain View', 'types': ['locality', 'political']}, {'long_name': 'Santa Clara County', 'short_name': 'Santa Clara County', 'types': ['administrative_area_level_2', 'political']}, {'long_name': 'California', 'short_name': 'CA', 'types': ['administrative_area_level_1', 'political']}, {'long_name': 'United States', 'short_name': 'US', 'types': ['country', 'political']}, {'long_name': '94043', 'short_name': '94043', 'types': ['postal_code']}], 'formatted_address': '1600 Amphitheatre Pkwy, Mountain View, CA 94043, USA', 'geometry': {'location': {'lat': 37.4213117, 'lng': -122.0839677}, 'location_type': 'ROOFTOP', 'viewport': {'northeast': {'lat': 37.4226606802915, 'lng': -122.0826187197085}, 'southwest': {'lat': 37.4199627197085, 'lng': -122.0853166802915

In [14]:
geocode_result[0]['geometry']['location']

{'lat': 37.4213117, 'lng': -122.0839677}

**Getting All Latitude and Longitude Values:**

In [15]:
def google_geocoder(address):
    ''' 
        Request geocode and returns Latitude and Longitude tuple.
    '''
    try:
        result = gmaps.geocode(address)
        latitude = result[0]['geometry']['location']['lat']
        longitude = result[0]['geometry']['location']['lng']
    except:
        latitude = np.nan
        longitude = np.nan
    return latitude, longitude

In [16]:
df['Coord'] = df['Location'].apply(google_geocoder)

In [17]:
print("Google Geocoding API was able to retrieve {} out of {} Boroughs".format(df['Coord'].count(), df.shape[0]))
df.head()

Google Geocoding API was able to retrieve 103 out of 103 Boroughs


Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

                                 Location  \
0        North York, M3A, Toronto, Canada   
1        North York, M4A, Toronto, Canada   
2  Downtown Toronto, M5A, Toronto, Canada   
3        North York, M6A, Toronto, Canada   
4  Downtown Toronto, M7A, Toronto, Canada   

                                           location2   Latitude  Longitude  \
0  (North York, Toronto, Golden Horseshoe, Ontari...  43.754326 -79.449117   
1                                               None        NaN        NaN   
2                                               None        NaN        NaN   
3                                               None        NaN        NaN   
4  (Downtown Yonge, Toronto Centre, Old Toronto, ...  43.656322 -79.380916   

                              Coord  
0         (43.7532586, -79.3296565)  
1  (43.72588229999999, -79.3155716)  
2  (43.6542599, -79.36063589999999)  
3          (43.718518, -79.4647633)  
4         (43.6623015, -79.3894938)

**Cleanup and format**

In [18]:
df['Latitude'] = df['Coord'].apply(lambda x: x[0])
df['Longitude'] = df['Coord'].apply(lambda x: x[1])

df = df.drop(columns=['Location', 'location2', 'Coord'])
df.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [19]:
# Saving file into Cloud Object Storage
project.save_data(file_name = "toronto_geolocation.csv", data = df.to_csv(index=False), overwrite=True)
print('Dataframe Saved as csv!')

Dataframe Saved as csv!


## <a name="explore"></a>Explore Neighborhoods in Toronto

**Visualize the Neighborhoods**

In [20]:
# create map of Toronto using Nominatim Toronto Latitude and Longitude

# get Toronto Geolocation coordinates
geolocator = Nominatim(user_agent='toronto_neigh')
location = geolocator.geocode('Toronto, Ontario, Canada')
latitude = location.latitude
longitude = location.longitude

# instantiate map
map_toronto = folium.Map(location=[latitude, longitude], 
                         zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
  label = '{}, {}'.format(neighborhood, borough)
  label = folium.Popup(label, parse_html=True)
  folium.CircleMarker(
      [lat, lng],
      radius=5,
      popup=label,
      color='blue',
      fill=True,
      fill_color='#3186cc',
      fill_opacity=0.7,
      parse_html=False).add_to(map_toronto)

map_toronto

## <a name="analize"></a>Analyze each Neighborhood

Define Foursquare Credentials and Version

In [21]:
# The code was removed by Watson Studio for sharing.

**Let's explore the first neighborhood in our dataset**

In [22]:
df.loc[0, 'Neighborhood']

'Parkwoods'

**Get the Neighborhood's Latitude and Longitude values:**

In [23]:
neighborhood_latitude = df.loc[0, 'Latitude'] # Neighborhood's Latitude
neighborhood_longitude = df.loc[0, 'Longitude'] # Neighborhood's Longitude
neighborhood_name = df.loc[0, 'Neighborhood'] # Neighborhood's Name

print("The Latitude and Longitude values of {} are {}, {}".format(neighborhood_name, neighborhood_latitude, neighborhood_longitude))

The Latitude and Longitude values of Parkwoods are 43.7532586, -79.3296565


**Now let's get the top 100 venues that are in proximity to this neighborhood**

In [24]:
limit = 100 # limit of number of venues returned by Foursquare API
radius = 500 
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        client_id, client_secret, version, neighborhood_latitude, neighborhood_longitude, radius, limit)
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e9ecaa3df2774001b5ed908'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

All information is in the Items key:

In [25]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we're ready to clean the json and structure it into a pandas dataframe

In [26]:
venues = results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

In [27]:
print('{} venues were returned by Foursquare'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare


### Explore all Neighborhoods in Toronto

Repeating same process with a function to explore all neighborhoods.

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            client_id, 
            client_secret, 
            version, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [35]:
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                latitudes=df['Latitude'],
                                longitudes=df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview
The Danforth West, Ri

Let's check the size of the resulting dataframe

In [36]:
print(toronto_venues.shape)
toronto_venues.head()

(2371, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

In [37]:
# Saving file into Cloud Object Storage
project.save_data(file_name = "toronto_venues.csv", data = toronto_venues.to_csv(index=False), overwrite=True)
print('Dataframe Saved as csv!')

Dataframe Saved as csv!


How many venues were found for each neighborhood:

In [39]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  9   
Bathurst Manor, Wilson Heights, Downsview North                        20   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           60   
Business reply mail Processing CentrE                                  17   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     65   
Christie                                                               18   
Church and Wellesley                                                   72   
Clarks Corners, Tam O'Shanter, Sullivan                                16   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       22   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              27   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           60   
East Toronto                                                            3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      9   
Fairview, Henry Farm, Oriole                                           67   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                               22   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                         8   
Guildwood, Morningside, West Hill                                       7   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          59   
Hillcrest Village                                                       5   
Humber Summit                                                           3   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         56   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              56   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       14   
Lawrence Park                                                           3   
Leaside                                                                34   
Little Portugal, Trinity                                               60   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

How many unique categories can be curated from all the returned venues:

In [40]:
print("There are {} unique categories.".format(len(toronto_venues['Venue Category'].unique())))

There are 276 unique categories.


### Analyze each Neighborhood

In [53]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0    0                  0        0                   0             0   
1    0                  0        0                   0             0   
2    0                  0        0                   0             0   
3    0                  0        0                   0             0   
4    0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Amphitheater  Antique Shop  Aquarium  Arcade  Art Gallery  Art Museum  \
0             0             0         0       0            0           0   
1             0             0         0       0            0           0   
2             0             0         0       0            0           0   
3             0             0         0       0            0           0   
4             0             0         0       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint 

In [58]:
col_list = toronto_onehot.columns.to_list()
col_list.remove('Neighborhood')
col_list.append('Neighborhood')
col_list[-1]
fixed_columns = [col_list[-1]] + col_list[:-1]
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhood  Zoo  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods    0                  0        0                   0   
1         Parkwoods    0                  0        0                   0   
2  Victoria Village    0                  0        0                   0   
3  Victoria Village    0                  0        0                   0   
4  Victoria Village    0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Amphitheater  Antique Shop  Aquarium  Arcade  \
0                    0             0             0         0       0   
1                    0             0             0         0       0   
2                    0             0             0         0       0   
3                    0             0             0         0       0   
4                    0             0             0         0       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0       0     0    0               0                 0                 0   
1       0     0    0               0                 0                 0   
2       0     0    0               0                 0                 0   
3       0     0    0               0                 0                 0   
4       0     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0          0       0              0          0         0   
1                   0          0       0              0          0         0   
2                   0          0       0              0          0         0   
3                   0          0       0              0          0         0   
4                   0          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  

**Let's group rows by Neighborhood and by taking the mean of the frequency of occurence of each category**

In [59]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood       Zoo  \
0                                           Agincourt  0.000000   
1                              Alderwood, Long Branch  0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North  0.000000   
3                                     Bayview Village  0.000000   
4                   Bedford Park, Lawrence Manor East  0.000000   
5                                         Berczy Park  0.000000   
6                         Birch Cliff, Cliffside West  0.000000   
7        Brockton, Parkdale Village, Exhibition Place  0.000000   
8               Business reply mail Processing CentrE  0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...  0.000000   
10                                Caledonia-Fairbanks  0.000000   
11              Canada Post Gateway Processing Centre  0.000000   
12                                          Cedarbrae  0.000000   
13                                 Central Bay Street  0.000000   
14                                           Christie  0.000000   
15                               Church and Wellesley  0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan  0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West  0.000000   
18                     Commerce Court, Victoria Hotel  0.000000   
19                                         Davisville  0.000000   
20                                   Davisville North  0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn  0.000000   
22                                          Don Mills  0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...  0.000000   
24                                          Downsview  0.000000   
25                       Dufferin, Dovercourt Village  0.000000   
26                                       East Toronto  0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...  0.000000   
28                       Fairview, Henry Farm, Oriole  0.000000   
29             First Canadian Place, Underground city  0.000000   
30                           Forest Hill North & West  0.000000   
31                           Garden District, Ryerson  0.000000   
32                                          Glencairn  0.000000   
33                    Golden Mile, Clairlea, Oakridge  0.000000   
34                  Guildwood, Morningside, West Hill  0.000000   
35  Harbourfront East, Union Station, Toronto Islands  0.000000   
36                      High Park, The Junction South  0.000000   
37                                  Hillcrest Village  0.000000   
38                                      Humber Summit  0.000000   
39                                   Humberlea, Emery  0.000000   
40                                 Humewood-Cedarvale  0.000000   
41                     India Bazaar, The Beaches West  0.000000   
42        Kennedy Park, Ionview, East Birchmount Park  0.000000   
43          Kensington Market, Chinatown, Grange Park  0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...  0.000000   
45                   Lawrence Manor, Lawrence Heights  0.000000   
46                                      Lawrence Park  0.000000   
47                                            Leaside  0.000000   
48                           Little Portugal, Trinity  0.000000   
49                                     Malvern, Rouge  0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...  0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...  0.000000   
52                        Moore Park, Summerhill East  0.000000   
53       New Toronto, Mimico South, Humber Bay Shores  0.000000   
54           North Park, Maple Leaf Park, Upwood Park  0.000000   
55                                 North Toronto West  0.000000   
56                                          Northwest  0.000000   
57                    Northwood Park, York University  0.000000   
58  Old Mill South, King's Mill Park, Sunnylea, Hu...  0.000000   
59                   

In [60]:
toronto_grouped.shape

(96, 276)

**Printing each Neighborhood along with the 5 most common venues**

In [61]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4                        Zoo  0.00


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.22
1                 Gym  0.11
2  Athletics & Sports  0.11
3        Skating Rink  0.11
4                 Pub  0.11


----Bathurst Manor, Wilson Heights, Downsview North----
              venue  freq
0              Bank  0.10
1       Coffee Shop  0.10
2          Pharmacy  0.05
3             Diner  0.05
4  Sushi Restaurant  0.05


----Bayview Village----
                       venue  freq
0                       Café  0.25
1         Chinese Restaurant  0.25
2                       Bank  0.25
3        Japanese Restaurant  0.25
4  Middle Eastern Restaurant  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0          Restaurant  0.09
1      Sandwich Place

In [62]:
# putting into Pandas Dataframe

# function to sort venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

# Dataframe to display top 10 venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1         Athletics & Sports                   Pub              Pharmacy   
2                       Bank              Pharmacy   Fried Chicken Joint   
3        Japanese Restaurant                  Bank    Chinese Restaurant   
4         Italian Restaurant            Restaurant           Coffee Shop   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant               Dog Run      Doner Restaurant   
1                          Gym           Coffee Shop          Skating Rink   
2                Shopping Mall           Bridal Shop        Sandwich Place   
3                  Yoga Studio             Drugstore               Dog Run   
4      Comfort Food Restaurant       Thai Restaurant              Boutique   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0            Donut Shop                  Drugstore    Dumpling Restaurant  
1        Sandwich Place         Dim Sum Restaurant                  Diner  
2                 Diner  Middle Eastern Restaurant            Supermarket  
3      Doner Restaurant                 Donut Shop    Dumpling Restaurant  
4             Juice Bar                    Butcher      Indian Restaurant

In [63]:
# Saving file into Cloud Object Storage
project.save_data(file_name = "toronto_neighborhoods_venues_sorted.csv", data = neighborhoods_venues_sorted.to_csv(index=False), overwrite=True)
print('Dataframe Saved as csv!')

Dataframe Saved as csv!


## <a name="cluster"></a>Cluster Neighborhoods

## <a name="examine"></a>Examine Clusters